extensions/amp-smartlinks/0.1/amp-smartlinks.js

```diff
@@ -0,0 +1,195 @@
+
+    /** @private {?../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = null;
+
+    /** @private {?../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = null;
+
+    /** @private {?./link-rewriter/link-rewriter-manager.LinkRewriterManager} */
+    this.linkRewriterService_ = null;
+
+    /** @private {?./link-rewriter/link-rewriter.LinkRewriter} */
+    this.smartLinkRewriter_ = null;
+
+    /** @private {?Object} */
+    this.linkmateOptions_ = null;
+
+    /** @private {?function({JsonObject})} */
+    this.fetchLinkmateConfig_ = null;
+
+    /** @private {?../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = null;
+
+    /** @private {?./linkmate.Linkmate} */
+    this.linkmate_ = null;
+
+    /** @private {?string} */
+    this.referrer_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.xhr_ = Services.xhrFor(this.win);
+    this.ampDoc_ = this.getAmpDoc();
+
+    this.viewer_ = Services.viewerForDoc(this.ampDoc_);
+
+    this.linkmateOptions_ = getConfigOptions(this.element);
+    this.fetchLinkmateConfig_ = this.getLinkmateOptions_;
+
+    this.linkRewriterService_ = new LinkRewriterManager(this.ampDoc_);
+
+    return this.ampDoc_.whenBodyAvailable()
+        .then(() => this.viewer_.getReferrerUrl())
+        .then(referrer => {
+          this.referrer_ = referrer;
+          this.runSmartlinks_();
+        });
+  }
+
+  /**
+   * Wait for the config promise to resolve and then proceed to functionality
+   * @private
+   */
+  runSmartlinks_() {
+    this.fetchLinkmateConfig_().then(config => {
+      this.linkmateOptions_.linkmateExpected = config['linkmate_enabled'];
+      this.linkmateOptions_.publisherID = config['publisher_id'];
+
+      this.postPageImpression_();
+      this.linkmate_ = new Linkmate(
+          this.ampDoc_,
+          this.xhr_,
+          this.linkmateOptions_
+      );
+    });
+
+    this.smartLinkRewriter_ = this.initLinkRewriter_();
+
+    // We have a valid linkmate class, the config specified linkmate to run
+    // and our API is expecting linkmate to run
+    if (this.linkmate_ && this.linkmateOptions_.linkmateEnabled &&
+      this.linkmateOptions_.linkmateExpected) {
+      this.smartLinkRewriter_.getAnchorReplacementList();
+    }
+  }
+
+  /**
+   * API call to retrieve the Narrativ config for this extension.
+   * @return {?Promise}
+   * @private
+   */
+  getLinkmateOptions_() {
+    const fetchUrl = ENDPOINTS.NRTV_CONFIG_ENDPOINT.replace(
+        '.nrtv_slug.', this.linkmateOptions_.nrtvSlug
+    );
+
+    return this.xhr_.fetchJson(fetchUrl, {
+      method: 'GET',
+      ampCors: false,
+    })
+        .then(res => res.json())
+        .then(res => {
+          return getData(res)[0]['amp_config'];
+        });
+  }
+
+  /**
+   * API call to indicate a page load event happened
+   * @private
+   */
+  postPageImpression_() {
+    const payload = this.buildPageImpressionPayload_();
+
+    this.xhr_.fetchJson(ENDPOINTS.PAGE_IMPRESSION_ENDPOINT, {
```

Comment:
+ @lannka Ideally I would recommend them to use our `CustomEventReporterBuilder`. However I don't believe we support sending request body with analytics custom event? Do you want to prioritize that feature and get it implemented? Thanks.

Reply:
chat offline. we do have body payload supported in `CustomEventReporterBuilder` as a `transport` level config.

Reply:
Would you like me to implement this API use `CustomEventReporterBuilder`? I think I can get it working from some examples in the codebase I looked at.

Reply:
@PhilWinchester Yes please migrate to the `CustomEventReporterBuilder` [API](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/amp-components-analytics.md#customeventreporter-api).  Let me know if there's any question. Thank you

Reply:
If I'm using `useBody: true` what's the best way to set the `extraUrlParams: { body: payload }`? For now I've created a helper function in `extension-analytics.js` similar to `setTransportConfig` but I'm not sure if that's the best way to go about doing this. 

```js
  postPageImpression_() {
    const builder = new CustomEventReporterBuilder(this.element);
    const payload = this.buildPageImpressionPayload_();

    builder.track('page-impression', ENDPOINTS.PAGE_IMPRESSION_ENDPOINT);
    builder.setTransportConfig(dict({
      'beacon': true,
      'xhrpost': false,
      'useBody': true,
    }));
    builder.setExtraUrlParams(payload);

    const reporter = builder.build();
    reporter.trigger('page-impression', dict());
  }
```

Reply:
@PhilWinchester That's a good point. I think adding a helper method to do that is fine for now. @lannka I know you wanted to use the `CustomEventReporter`. I do see a need to redesign the API a bit because there're many new features added to amp-analytics. If so we can create a helper function to unblock the pr for now.